### ISSUE TITLE
Branches with same content does not show empty diff

### ISSUE THREAD

**Issue by jedi7** [go-gitea/gitea#19666]
### Description

When user creates same content in separate branches. Then the diff shows differences even if the tree content is the same.

See repo https://try.gitea.io/jedi7/test_paralel_branches/src/branch/paralel

Also please see the screenshots

### Gitea Version

1.16.7

### Can you reproduce the bug on the Gitea demo site?

Yes

### Screenshots

Master branch with "test" content
![image](https://user-images.githubusercontent.com/24191306/167545261-dc9091da-ecd3-47cd-a17f-4fc8073c1bfe.png)

Paralel branch with "test" content (added in different commit, so conflicting with master)
![image](https://user-images.githubusercontent.com/24191306/167545346-1604103a-8b62-45c6-b30f-43858dd6c38e.png)

Invalid diff
![image](https://user-images.githubusercontent.com/24191306/167545460-33ad2a64-8a08-44f7-bead-1c17c100220e.png)

Invalid PR
![image](https://user-images.githubusercontent.com/24191306/167545563-98ff19e7-91ee-4d59-aaf7-00af92e5d72f.png)

### Git Version

web

### Operating System

Archlinux

### How are you running Gitea?

https://try.gitea.io/ 
also reproducible on released gitea 1.16.7

### Database

SQLite

**Comment by Gusted** [comment]
> Then the diff shows differences even if the tree content is the same.

The diff is generated based of the older base branch commit. This diff seems to be as intended.

**Comment by jedi7** [comment]
> The diff is generated based of the older base branch commit. This diff seems to be as intended.
But why it should be intended? When I compare two branches, I'm interested in actual state and not what was before X commits.

**Comment by Gusted** [comment]
> When comparing two branches, then user is interested in actual state (not what was before X commits).

In such case they should update the branch to include the latest change from the base branch. I'm pretty sure this is the same behavior for GitHub+GitLab.

**Comment by jedi7** [comment]
@Gusted  true, github have the same behavior (just tested on the testing repo)
But still not understand how it is useful.
when comparing "master" and "paralel" branches. 

master content: "test"
paralel content: "test"
github/gitea diff is "test"

but diff should be ""  or at least say the content "test" is conflicting.
when I do `git diff master paralel` 
[test-empty-merge-commit-paralel.tar.gz](https://github.com/go-gitea/gitea/files/8667138/test-empty-merge-commit-paralel.tar.gz)
, then diff is empty.

Attaching test repo.

**Comment by Gusted** [comment]
> when I do `git diff master paralel` , then diff is empty.

Gitea doesn't use that diff, under-the-hood Gitea requests [the merge-base](https://git-scm.com/docs/git-merge-base) of the `paralel` with `master` branch: https://github.com/go-gitea/gitea/blob/main/modules/git/repo_compare.go#L35 and that's the actual commit being used to diff against. In your case the command would return `df1d83b5339dfe1155a12e0cc86085e3432addf1` which is the initial commit.

As of why we do this? I'm not entirely sure, but I assume if we would diff against the branch directly, any kind of conflict between the branches would lead to a inaccurate presentation of the diff.

**Comment by jedi7** [comment]
Thank you for explanation of the mechanics in gitea, I think I understand now.
Question is, if this behavior is ok, or is there space for improve.

From my point of view will be useful this behavior:
* diff actual branches and mark what was added, removed and (newly) what is conflicting.

Actual behavior is for me little misleading, because for example I can think for I'm fixing a bug, but somebody else did it already and I will not see it in diff.

**Comment by hramrach** [comment]
It's common but not universal practice to only merge up-to-date branches.

The gitea UI clearly does not support the other option - merging branches that are not up-to-date.

Some forges provide a 'rebase' button that does automated rebase.

Many forges also show if the branch can be merged or if it conflicts in the PR status, and that's a useful feature to have.

To calculate if there is a conflict the merge needs to be done, and then diff of the merge against the branch head can be shown as well.

**Comment by Eddcapone** [comment]
I was confused too. It would be nice if the "Changed Files" Tab would only show the files, which actually changed.
I suggest that you provide another diff option so that only the actual changed files are showing.

You could run a foreach on all files and execute a `git diff target feature $filename` and if the output is empty, then simply don't show the file.

**Comment by lunny** [comment]
This is a normal behavior which is the same as Github
